### PR TITLE
Make NF4Tensor constructible from different dtypes

### PR DESCRIPTION
**Pull request by drisspg**
# Summary
This doesnt work today because transpose is doing some dtype shennagins

### CODE REVIEW

**Comment by drisspg** [comment]
Failures
```Shell
========================================================================================================= short test summary info =========================================================================================================
FAILED test/dtypes/test_nf4.py::TestNF4Linear::test_output_dtype_match[dtype0] - RuntimeError: "copy_" not implemented for 'Bits2x4'
FAILED test/dtypes/test_nf4.py::TestNF4Linear::test_backward_dtype_match[dtype0] - RuntimeError: "copy_" not implemented for 'Bits2x4'
FAILED test/dtypes/test_nf4.py::TestNF4Linear::test_nf4_bnb_linear[dtype0] - RuntimeError: copy_() does not support casting bits types to different bits types. Source dtype is BFloat16target dtype is Bits2x4
FAILED test/dtypes/test_nf4.py::TestNF4Linear::test_to_bfloat16 - AssertionError: assert <class 'torchao.dtypes.nf4tensor.NF4Tensor'> == <class 'torch.Tensor'>
FAILED test/dtypes/test_nf4.py::TestNF4Linear::test_smoketest_linear[dtype0] - RuntimeError: copy_() does not support casting bits types to different bits types. Source dtype is BFloat16target dtype is Bits2x4
FAILED test/dtypes/test_nf4.py::TestNF4Linear::test_smoketest_linear[dtype1] - RuntimeError: copy_() does not support casting bits types to different bits types. Source dtype is BFloat16target dtype is Bits2x4
FAILED test/dtypes/test_nf4.py::TestNF4Linear::test_smoketest_linear[dtype2] - RuntimeError: copy_() does not support casting bits types to different bits types. Source dtype is BFloat16target dtype is Bits2x4
FAILED test/dtypes/test_nf4.py::TestNF4Linear::test_smoketest_linear_compile[dtype0] - torch._dynamo.exc.TorchRuntimeError: Failed running call_function <built-in function linear>(*(FakeTensor(..., device='cuda:0', size=(2, 32, 32), dtype=torch.bfloat16), Quantized Data: FakeTensor(..., device='cuda:0', size=(512,),...
FAILED test/dtypes/test_nf4.py::TestNF4Linear::test_smoketest_linear_compile[dtype1] - torch._dynamo.exc.TorchRuntimeError: Failed running call_function <built-in function linear>(*(FakeTensor(..., device='cuda:0', size=(2, 32, 32), dtype=torch.float16), Quantized Data: FakeTensor(..., device='cuda:0', size=(512,), ...
FAILED test/dtypes/test_nf4.py::TestNF4Linear::test_smoketest_linear_compile[dtype2] - torch._dynamo.exc.TorchRuntimeError: Failed running call_function <built-in function linear>(*(FakeTensor(..., device='cuda:0', size=(2, 32, 32)), Quantized Data: FakeTensor(..., device='cuda:0', size=(512,), dtype=torch.uint8)
================================================================================================ 10 failed, 22 passed, 3 warnings in 9.22s ================================================================================================
```

File: torchao/dtypes/nf4tensor.py [hunk 1]
```diff
@@ -186,8 +186,7 @@ def __new__(
             tensor_meta.original_shape,
             tensor_meta.original_strides,
             tensor_meta.storage_offset,
-            # Picked some floating dtype, but we need dtype extensibility
-            dtype=torch.float8_e5m2fnuz,
+            dtype=tensor_meta.dtype, # Dtype broadcast to the world matches input
```
Comment (cpuhrsch) [inline torchao/dtypes/nf4tensor.py:189]: Input during construction. Does this mean `.to(torch.float32)` should change this `dtype` as well and also cause different results when someone calls `dequantize` or somesuch?

Comment (cpuhrsch) [reply torchao/dtypes/nf4tensor.py:189]: It does make sense to have this to figure out the dtype of the lookup table and better cache the dequantize operation (which could also be mitigated with multiple lookup tables or kernels).

**Comment by cpuhrsch** [comment]
Thanks for sending this!

File: test/dtypes/test_nf4.py [hunk 2]
```diff
@@ -57,42 +58,46 @@ def save_state_dict_to_buffer(self, state_dict: OrderedDict):
         buffer.seek(0)
         return buffer
 
-    def test_register_nf4_as_param(self):
-        nf4_tensor = to_nf4(torch.randn(512, 512, dtype=torch.bfloat16))
+    @pytest.mark.parametrize("dtype", [torch.bfloat16, torch.float16, torch.float32])
+    def test_register_nf4_as_param(self, dtype: torch.dtype):
+        nf4_tensor = to_nf4(torch.randn(512, 512, dtype=dtype))
```
Comment (weifengpy) [inline test/dtypes/test_nf4.py:63]: do we need to remove bf16 in `to_nf4` as well?
```
def to_nf4(tensor, block_size: int = 64, scaler_block_size: int = 256):
    tensor1 = tensor.to(torch.bfloat16)
    return NF4Tensor.from_tensor(tensor1, block_size, scaler_block_size)
```

Comment (weifengpy) [reply test/dtypes/test_nf4.py:63]: to make torchtune fp32 work, I reuse this unit test directly: https://github.com/pytorch-labs/ao/pull/118

this one is passing now in my PR. there are still gaps in state dict and nn.functional

Comment (drisspg) [reply test/dtypes/test_nf4.py:63]: great catch!

**Pull request closed without merging**